### PR TITLE
feat(cheap-dapp): Initial version

### PR DESCRIPTION
**Pull request by qbzzt**
Closing: INT-596

### CODE REVIEW

File: src/docs/developers/build/cheap-dapp.md [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+In almost all cases, the L1 data fee is the vast majority of the transaction's cost, and the L2 execution fee is, comparatively speaking, negligible.
+This means that the optimization tradeoffs are very different in Optimism than they are in Ethereum.
+
+Transaction call data is *expensive*.
+The cost of writing a byte to L1 is approximately 16 gas.
+As I'm writing this, this means 720 gwei, or 720,000 units of L2 gas.
+
+In comparison, on-chain processing and storage are cheap.
+The worst case for writing to storage (previously uninitialized storage) is a cost of [22100 gas per 32 byte word](https://www.evm.codes/#55), which averages out to less than 700 gas / byte.
```
Comment (Nicca42) [inline src/docs/developers/build/cheap-dapp.md:39]: Is it a word of byte? Is it not per 32 byte string?

Comment (Nicca42) [reply src/docs/developers/build/cheap-dapp.md:39]: Also as we are jumping between L1 and L2 gas here, please specify when you say gas which gas you are referring to. 

> The worst case for writing to storage ~~(previously uninitialized storage)~~ is a cost of [22100 L2 gas per 32 byte word]

File: src/docs/developers/build/cheap-dapp.md [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+At writing, this means it is cheaper to write a whole kilobyte to storage, rather than add one byte to the transaction call data.
```
Comment (Nicca42) [inline src/docs/developers/build/cheap-dapp.md:40]: > ~~At writing,~~ This means it is cheaper...

File: src/docs/developers/build/cheap-dapp.md [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+Transaction call data is *expensive*.
+The cost of writing a byte to L1 is approximately 16 gas.
+As I'm writing this, this means 720 gwei, or 720,000 units of L2 gas.
+
+In comparison, on-chain processing and storage are cheap.
+The worst case for writing to storage (previously uninitialized storage) is a cost of [22100 gas per 32 byte word](https://www.evm.codes/#55), which averages out to less than 700 gas / byte.
+At writing, this means it is cheaper to write a whole kilobyte to storage, rather than add one byte to the transaction call data. 
+
+## Modify the ABI (application binary interface)
```
Comment (Nicca42) [inline src/docs/developers/build/cheap-dapp.md:42]: > Modify the ABI (Application Binary Interface)

Also link [ABI](https://docs.soliditylang.org/en/develop/abi-spec.html) to definition.

File: src/docs/developers/build/cheap-dapp.md [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+[The standard ABI](https://docs.soliditylang.org/en/latest/abi-spec.html) was designed with L1 tradeoffs in mind. 
+It uses four byte function selectors and pads values into 32 byte words. 
+Neither is optimal when using Optimism.
+
+[It is much more efficient to create a shorter ABI with just the required bytes, and decode it onchain](https://ethereum.org/en/developers/tutorials/short-abi/).
```
Comment (Nicca42) [inline src/docs/developers/build/cheap-dapp.md:48]: > It is much more efficient to [create a shorter ABI with just the required bytes, and decode it onchain]()

File: src/docs/developers/build/cheap-dapp.md [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+To get better compatibility, create two contracts:
```
Comment (Nicca42) [inline src/docs/developers/build/cheap-dapp.md:49]: This feels too specific to the docs, and quite hard to follow in this form. Can we move this to a tutorial? 

Remove everything between 49 and 55, leaving 56: 

> All of your ...

And 48:

> It is much more ...

File: src/docs/developers/build/cheap-dapp.md [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+Note that only transactions are written to L1.
+All of your [`view`](https://docs.soliditylang.org/en/latest/contracts.html#view-functions) and [`pure`](https://docs.soliditylang.org/en/latest/contracts.html#pure-functions) functions can use the standard ABI at no cost.
+
+
+## Use smaller values when possible
+
+Your modified ABI is not going to pad values, so the less bytes you use the better.
+For example, it is standard to use `uint256` for amounts.
+This means that the highest amount of ETH we can represent is 2<sup>256</sup>-1 wei, or about 1.2*10<sup>59</sup> ETH.
```
Comment (Nicca42) [inline src/docs/developers/build/cheap-dapp.md:63]: Should not be talking about ETH and amounts in the same sentence. 

> For example, it is standard to use `uint256` to store values. This means the highest value you can store is 2<sup>256</sup>-1, or about 1.2*10<sup>59</sup>.

File: src/docs/developers/build/cheap-dapp.md [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+Note that only transactions are written to L1.
+All of your [`view`](https://docs.soliditylang.org/en/latest/contracts.html#view-functions) and [`pure`](https://docs.soliditylang.org/en/latest/contracts.html#pure-functions) functions can use the standard ABI at no cost.
+
+
+## Use smaller values when possible
+
+Your modified ABI is not going to pad values, so the less bytes you use the better.
+For example, it is standard to use `uint256` for amounts.
+This means that the highest amount of ETH we can represent is 2<sup>256</sup>-1 wei, or about 1.2*10<sup>59</sup> ETH. 
+With a total amount of minted ETH around [120 million](https://ycharts.com/indicators/ethereum_supply), this is overkill.
```
Comment (Nicca42) [inline src/docs/developers/build/cheap-dapp.md:64]: When storing ETH balances using `uint256` is overkill as there is only [120 million ETH](). Thus, we can safely store ETH balances in `uint88` which is just eleven bytes...

File: src/docs/developers/build/cheap-dapp.md [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+
+To reduce costs even further, do we really need to be able to specify values to the last wei? 
+As I'm writing this, it would take approximately 4000 gwei (or 4*10<sup>21</sup> wei) to make one US cent ([see current values here](https://nomics.com/markets/gwei-gwei/usd-united-states-dollar)). 
+If we specify our amounts in 100 gwei units, and use only six bytes (`uint48`), we can still represent any value below 28 million ETH at an accuracy of a tenth of a cent.
+Just remember to rescale by deploying new code when ETH reaches $100,000 and a 100 gwei is equal to a cent.
+
+
+## Cache as cache can
+
```
Comment (Nicca42) [inline src/docs/developers/build/cheap-dapp.md:75]: I do not follow this at all. Are you saying to "cache" in the contracts? Because then you should say store it, not cache. Unless you mean cacheing on the front end? Generally not following.

**Review by Nicca42 — CHANGES_REQUESTED** [review]

File: src/docs/developers/build/cheap-dapp.md (hunk 6 above)
Comment (qbzzt) [reply src/docs/developers/build/cheap-dapp.md:63]: OK. However, I'm not using *store* on purpose. What we're trying to do is reduce the cost of calldata, which is essentially a transmission mechanism.

BTW, 2^256-1 is about 1.2*10^77. The 59 is because of changing from wei to ETH.

File: src/docs/developers/build/cheap-dapp.md (hunk 5 above)
Comment (qbzzt) [reply src/docs/developers/build/cheap-dapp.md:49]: I already wrote an article for ethereum.org about this, so I'm reluctant to also write a tutorial that will cover mostly the same ground. But if you think this is too specific we can just remove it here.

File: src/docs/developers/build/cheap-dapp.md [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+[See here for additional details about how transaction fees are calculated](transaction-fees.md).
+
+### Optimization tradeoffs
+
+In almost all cases, the L1 data fee is the vast majority of the transaction's cost, and the L2 execution fee is, comparatively speaking, negligible.
+This means that the optimization tradeoffs are very different in Optimism than they are in Ethereum.
+
+Transaction call data is *expensive*.
+The cost of writing a byte to L1 is approximately 16 gas.
+As I'm writing this, this means 720 gwei, or 720,000 units of L2 gas.
```
[the remark this answers is not in the document]
Comment (Nicca42) [reply src/docs/developers/build/cheap-dapp.md:41]: Please update this

File: src/docs/developers/build/cheap-dapp.md (hunk 1 above)
Comment (Nicca42) [reply src/docs/developers/build/cheap-dapp.md:44]: ^

File: src/docs/developers/build/cheap-dapp.md (hunk 2 above)
Comment (Nicca42) [reply src/docs/developers/build/cheap-dapp.md:45]: ^

File: src/docs/developers/build/cheap-dapp.md (hunk 4 above)
Comment (Nicca42) [reply src/docs/developers/build/cheap-dapp.md:53]: ^

**Review by Nicca42 — CHANGES_REQUESTED** [review]

File: src/docs/developers/build/cheap-dapp.md (hunk 1 above)
Comment (qbzzt) [reply src/docs/developers/build/cheap-dapp.md:44]: > Is it a word of byte? Is it not per 32 byte string?

An EVM word is defined as 256 bits, a.k.a. 32 bytes. I'll make that clearer.

File: src/docs/developers/build/cheap-dapp.md [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+
+Go through your contracts and identify any values that will never reach 32 bytes and reduce them to logical sizes. You can do this same process for ints, bytes and [other Solidity data types](https://docs.soliditylang.org/en/develop/types.html#types).
+
+## Store anything you can
+
+Call data is expensive on Optimism (relatively speaking), but storage is very cheap ([see above](#optimization-tradeoffs)).
+So any value that is likely to be used multiple times should be stored on chain if possible.
+If you are writing your own ABI, you can use variable length fields for those references.
+
+For example, when specifying an address in a call from your client to your contract, you can specify that if the value starts with 0x00 the rest of the value is the twenty bytes of an uncached address. If the value starts ix 0x1 then the next 12 bits (the last four bitys of the first value followed by one byte) are a storage key. If you need to store more information, 0x2 can mark a 20 bit cache key (the last four bits followed by two bytes), etc.
```
[the remark this answers is not in the document]
Comment (qbzzt) [reply src/docs/developers/build/cheap-dapp.md:72]: To properly explain this would take an article. I'm going to remove this for now, and write the article for ethereum.org, ok?

[the remark this answers is not in the document]
Comment (Nicca42) [reply src/docs/developers/build/cheap-dapp.md]: Works for me

**Review by Nicca42 — APPROVED** [review]
